File: typescript/src/password.ts
import { decodeBase64Url, utf8 } from "./encoding.js";
import type { KeyDerivationParameters } from "./model.js";
import { keyHoldError } from "./errors.js";
import {
  KDF_ALGORITHM,
  MAX_ITERATIONS,
  MIN_ITERATIONS,
  OUTPUT_LENGTH_BITS,
  PASSWORD_ENCODING,
} from "./constants.js";

export async function derivePassword(
  password: string,
  salt: Uint8Array,
  parameters: KeyDerivationParameters,
): Promise<Uint8Array> {
  if (password.length === 0)
    throw keyHoldError("invalid_parameter", "password must not be empty");
  const passwordBytes = utf8(password);
  if (
    parameters.algorithm !== KDF_ALGORITHM ||
    parameters.passwordEncoding !== PASSWORD_ENCODING ||
    parameters.outputLengthBits !== OUTPUT_LENGTH_BITS ||
    !Number.isSafeInteger(parameters.iterations) ||
    parameters.iterations < MIN_ITERATIONS ||
    parameters.iterations > MAX_ITERATIONS
  )
    throw keyHoldError(
      "invalid_parameter",
      "invalid key derivation parameters",
    );
  const key = await globalThis.crypto.subtle.importKey(
    "raw",
    passwordBytes as unknown as BufferSource,
    "PBKDF2",
    false,
    ["deriveBits"],
  );
  return new Uint8Array(
    await globalThis.crypto.subtle.deriveBits(
      {
        name: "PBKDF2",
        salt: salt as unknown as BufferSource,
        iterations: parameters.iterations,
        hash: "SHA-256",
      },
      key,
      OUTPUT_LENGTH_BITS,
    ),
  );
}

export function saltFromDocument(value: string): Uint8Array {
  return decodeBase64Url(value, "keyDerivation.saltB64Url");
}

File: go/password.go
package keyhold

import (
	"crypto/sha256"
	"golang.org/x/crypto/pbkdf2"
	"unicode/utf8"
)

func derivePassword(password string, salt []byte, iterations int64) []byte {
	return pbkdf2.Key([]byte(password), salt, int(iterations), OutputLengthBits/8, sha256.New)
}
func validateParameters(parameters Parameters) error {
	if parameters.KeyDerivation.Algorithm != KDFAlgorithm || parameters.KeyDerivation.PasswordEncoding != PasswordEncoding || parameters.KeyDerivation.OutputLengthBits != OutputLengthBits || parameters.KeyDerivation.Iterations < MinIterations || parameters.KeyDerivation.Iterations > MaxIterations {
		return E(ErrInvalidParameter, "invalid key derivation parameters", nil)
	}
	if parameters.Cipher.Algorithm != CipherAlgorithm || parameters.Cipher.KeyLengthBits != KeyLengthBits || parameters.Cipher.TagLengthBits != TagLengthBits {
		return E(ErrInvalidParameter, "invalid cipher parameters", nil)
	}
	return nil
}
func validateText(label, password string) error {
	if !utf8.ValidString(label) || label == "" {
		return E(ErrInvalidParameter, "label must be a non-empty Unicode scalar sequence", nil)
	}
	if password == "" || !utf8.ValidString(password) {
		return E(ErrInvalidParameter, "password must be a non-empty Unicode scalar sequence", nil)
	}
	return nil
}
